File: app/routes/calendar.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, extract
from datetime import datetime, date
from typing import List, Optional
from calendar import monthrange
import calendar

from app.database import get_db
from app import models, schemas, auth

router = APIRouter()

@router.get("/admin", response_model=List[schemas.CalendarEvent])
async def get_admin_calendar(
    year: int = Query(..., description="Année à afficher"),
    month: int = Query(..., description="Mois à afficher (1-12)"),
    current_user: models.User = Depends(auth.get_current_admin_user),
    db: Session = Depends(get_db)
):
    """
    Récupère les événements du calendrier pour l'admin (vue mensuelle)
    Affiche toutes les demandes d'absence de tous les utilisateurs
    """
    # Calculer les dates de début et fin du mois
    start_date = date(year, month, 1)
    _, last_day = monthrange(year, month)
    end_date = date(year, month, last_day)
    
    # Récupérer toutes les demandes d'absence du mois
    absence_requests = db.query(models.AbsenceRequest).join(
        models.User, models.AbsenceRequest.user_id == models.User.id
    ).filter(
        and_(
            or_(
                # Demandes qui commencent dans le mois
                and_(
                    models.AbsenceRequest.start_date >= start_date,
                    models.AbsenceRequest.start_date <= end_date
                ),
                # Demandes qui finissent dans le mois
                and_(
                    models.AbsenceRequest.end_date >= start_date,
                    models.AbsenceRequest.end_date <= end_date
                ),
                # Demandes qui couvrent tout le mois
                and_(
                    models.AbsenceRequest.start_date <= start_date,
                    models.AbsenceRequest.end_date >= end_date
                )
            ),
            models.User.is_active == True
        )
    ).all()
    
    # Récupérer aussi les déclarations de maladie du mois
    sickness_declarations = db.query(models.SicknessDeclaration).join(
        models.User, models.SicknessDeclaration.user_id == models.User.id
    ).filter(
        and_(
            or_(
                # Déclarations qui commencent dans le mois
                and_(
                    models.SicknessDeclaration.start_date >= start_date,
                    models.SicknessDeclaration.start_date <= end_date
                ),
                # Déclarations qui finissent dans le mois
                and_(
                    models.SicknessDeclaration.end_date >= start_date,
                    models.SicknessDeclaration.end_date <= end_date
                ),
                # Déclarations qui couvrent tout le mois
                and_(
                    models.SicknessDeclaration.start_date <= start_date,
                    models.SicknessDeclaration.end_date >= end_date
                )
            ),
            models.User.is_active == True
        )
    ).all()
    
    # Convertir en événements calendrier
    calendar_events = []
    
    # Traiter les demandes d'absence
    for request in absence_requests:
        # Créer le titre avec nom de l'utilisateur et type
        type_label = "Vacances" if request.type == models.AbsenceType.VACANCES else "Maladie"
        status_label = ""
        if request.status == models.AbsenceStatus.EN_ATTENTE:
            status_label = " (En attente)"
        elif request.status == models.AbsenceStatus.REFUSE:
            status_label = " (Refusé)"
            
        title = f"{request.user.first_name} {request.user.last_name} - {type_label}{status_label}"
        
        calendar_events.append(schemas.CalendarEvent(
            id=request.id,
            title=title,
            start=max(request.start_date, start_date),  # Ajuster au début du mois si nécessaire
            end=min(request.end_date, end_date),        # Ajuster à la fin du mois si nécessaire
            type=request.type,
            status=request.status,
            user_name=f"{request.user.first_name} {request.user.last_name}",
            reason=request.reason,
            event_source="absence_request"
        ))
    
    # Traiter les déclarations de maladie
    for declaration in sickness_declarations:
        # Créer le titre avec nom de l'utilisateur
        email_status = " ✉️" if declaration.email_sent else " ❌"
        title = f"{declaration.user.first_name} {declaration.user.last_name} - Arrêt maladie{email_status}"
        
        # Les déclarations de maladie sont toujours de type MALADIE et approuvées
        calendar_events.append(schemas.CalendarEvent(
            id=declaration.id,
            title=title,
            start=max(declaration.start_date, start_date),
            end=min(declaration.end_date, end_date),
            type=models.AbsenceType.MALADIE,
            status=models.AbsenceStatus.APPROUVE,  # Les déclarations sont automatiquement approuvées
            user_name=f"{declaration.user.first_name} {declaration.user.last_name}",
            reason=(declaration.pdf_filename or declaration.description),
            event_source="sickness_declaration"
        ))
    
    return calendar_events

@router.get("/user", response_model=List[schemas.CalendarEvent])
async def get_user_calendar(
    year: int = Query(..., description="Année à afficher"),
    current_user: models.User = Depends(auth.get_current_active_user),
    db: Session = Depends(get_db)
):
    """
    Récupère les événements du calendrier pour un utilisateur (vue annuelle)
    Affiche uniquement ses propres demandes d'absence
    """
    # Calculer les dates de début et fin de l'année
    start_date = date(year, 1, 1)
    end_date = date(year, 12, 31)
    
    # Récupérer les demandes d'absence de l'utilisateur pour l'année
    absence_requests = db.query(models.AbsenceRequest).filter(
        and_(
            models.AbsenceRequest.user_id == current_user.id,
            or_(
                # Demandes qui commencent dans l'année
                and_(
                    models.AbsenceRequest.start_date >= start_date,
                    models.AbsenceRequest.start_date <= end_date
                ),
                # Demandes qui finissent dans l'année
                and_(
                    models.AbsenceRequest.end_date >= start_date,
                    models.AbsenceRequest.end_date <= end_date
                ),
                # Demandes qui couvrent toute l'année
                and_(
                    models.AbsenceRequest.start_date <= start_date,
                    models.AbsenceRequest.end_date >= end_date
                )
            )
        )
    ).all()
    
    # Récupérer aussi les déclarations de maladie de l'utilisateur
    sickness_declarations = db.query(models.SicknessDeclaration).filter(
        and_(
            models.SicknessDeclaration.user_id == current_user.id,
            or_(
                # Déclarations qui commencent dans l'année
                and_(
                    models.SicknessDeclaration.start_date >= start_date,
                    models.SicknessDeclaration.start_date <= end_date
                ),
                # Déclarations qui finissent dans l'année
                and_(
                    models.SicknessDeclaration.end_date >= start_date,
                    models.SicknessDeclaration.end_date <= end_date
                ),
                # Déclarations qui couvrent toute l'année
                and_(
                    models.SicknessDeclaration.start_date <= start_date,
                    models.SicknessDeclaration.end_date >= end_date
                )
            )
        )
    ).all()
    
    # Convertir en événements calendrier
    calendar_events = []
    
    # Traiter les demandes d'absence
    for request in absence_requests:
        type_label = "Vacances" if request.type == models.AbsenceType.VACANCES else "Maladie"
        status_label = ""
        if request.status == models.AbsenceStatus.EN_ATTENTE:
            status_label = " (En attente)"
        elif request.status == models.AbsenceStatus.REFUSE:
            status_label = " (Refusé)"
            
        title = f"{type_label}{status_label}"
        
        calendar_events.append(schemas.CalendarEvent(
            id=request.id,
            title=title,
            start=max(request.start_date, start_date),
            end=min(request.end_date, end_date),
            type=request.type,
            status=request.status,
            user_name=f"{current_user.first_name} {current_user.last_name}",
            reason=request.reason,
            event_source="absence_request"
        ))
    
    # Traiter les déclarations de maladie
    for declaration in sickness_declarations:
        email_status = " ✉️" if declaration.email_sent else " ❌"
        title = f"Arrêt maladie{email_status}"
        
        calendar_events.append(schemas.CalendarEvent(
            id=declaration.id,
            title=title,
            start=max(declaration.start_date, start_date),
            end=min(declaration.end_date, end_date),
            type=models.AbsenceType.MALADIE,
            status=models.AbsenceStatus.APPROUVE,
            user_name=f"{current_user.first_name} {current_user.last_name}",
            reason=declaration.description,
            event_source="sickness_declaration"
        ))
    
    return calendar_events

@router.get("/summary")
async def get_calendar_summary(
    year: int = Query(..., description="Année pour le résumé"),
    current_user: models.User = Depends(auth.get_current_active_user),
    db: Session = Depends(get_db)
):
    """
    Récupère un résumé des congés pour l'année donnée
    """
    # Calculer les jours utilisés dans l'année
    start_date = date(year, 1, 1)
    end_date = date(year, 12, 31)
    
    approved_requests = db.query(models.AbsenceRequest).filter(
        and_(
            models.AbsenceRequest.user_id == current_user.id,
            models.AbsenceRequest.status == models.AbsenceStatus.APPROUVE,
            models.AbsenceRequest.type == models.AbsenceType.VACANCES,
            models.AbsenceRequest.start_date <= end_date,
            models.AbsenceRequest.end_date >= start_date
        )
    ).all()
    
    # Calculer les jours utilisés
    used_days = 0
    for request in approved_requests:
        # Ajuster les dates pour rester dans l'année
        request_start = max(request.start_date, start_date)
        request_end = min(request.end_date, end_date)
        
        # Calculer le nombre de jours (inclus)
        days_count = (request_end - request_start).days + 1
        used_days += days_count
    
    remaining_days = max(0, current_user.annual_leave_days - used_days)
    
    return {
        "year": year,
        "total_leave_days": current_user.annual_leave_days,
        "used_leave_days": used_days,
        "remaining_leave_days": remaining_days
    }

@router.get("/summary/user/{user_id}")
async def get_calendar_summary_for_user(
    user_id: int,
    year: int = Query(..., description="Année pour le résumé"),
    current_user: models.User = Depends(auth.get_current_admin_user),
    db: Session = Depends(get_db)
):
    """Résumé des congés pour un utilisateur spécifique (admin)."""
    # Calculer les jours utilisés dans l'année pour l'utilisateur ciblé
    start_date = date(year, 1, 1)
    end_date = date(year, 12, 31)

    user = db.query(models.User).filter(models.User.id == user_id).first()
    if not user:
        raise HTTPException(status_code=404, detail="Utilisateur non trouvé")

    approved_requests = db.query(models.AbsenceRequest).filter(
        and_(
            models.AbsenceRequest.user_id == user_id,
            models.AbsenceRequest.status == models.AbsenceStatus.APPROUVE,
            models.AbsenceRequest.type == models.AbsenceType.VACANCES,
            models.AbsenceRequest.start_date <= end_date,
            models.AbsenceRequest.end_date >= start_date
        )
    ).all()

    used_days = 0
    for request in approved_requests:
        request_start = max(request.start_date, start_date)
        request_end = min(request.end_date, end_date)
        days_count = (request_end - request_start).days + 1
        used_days += days_count

    remaining_days = max(0, user.annual_leave_days - used_days)

    return {
        "year": year,
        "total_leave_days": user.annual_leave_days,
        "used_leave_days": used_days,
        "remaining_leave_days": remaining_days
    }

File: static/js/calendar.js
// Gestion du calendrier (vue mensuelle admin, vue annuelle utilisateur)

class Calendar {
    constructor() {
        this.currentDate = new Date();
        this.currentYear = this.currentDate.getFullYear();
        this.currentMonth = this.currentDate.getMonth();
        this.isAdmin = false;
        this.events = [];
        this.monthNames = [
            'Janvier', 'Février', 'Mars', 'Avril', 'Mai', 'Juin',
            'Juillet', 'Août', 'Septembre', 'Octobre', 'Novembre', 'Décembre'
        ];
    }

    async init() {
        this.isAdmin = currentUser && currentUser.role === 'admin';
        
        this.setupEventListeners();
        await this.showCalendar();
    }

    setupEventListeners() {
        const prevBtn = document.getElementById('prev-period');
        const nextBtn = document.getElementById('next-period');
        const todayBtn = document.getElementById('today-btn');
        
        if (prevBtn) prevBtn.addEventListener('click', () => this.navigatePrevious());
        if (nextBtn) nextBtn.addEventListener('click', () => this.navigateNext());
        if (todayBtn) todayBtn.addEventListener('click', () => this.goToToday());

    }

    navigatePrevious() {
        if (this.isAdmin) {
            this.currentMonth--;
            if (this.currentMonth < 0) {
                this.currentMonth = 11;
                this.currentYear--;
            }
        } else {
            this.currentYear--;
        }
        this.showCalendar();
    }

    navigateNext() {
        if (this.isAdmin) {
            this.currentMonth++;
            if (this.currentMonth > 11) {
                this.currentMonth = 0;
                this.currentYear++;
            }
        } else {
            this.currentYear++;
        }
        this.showCalendar();
    }

    goToToday() {
        const today = new Date();
        this.currentYear = today.getFullYear();
        this.currentMonth = today.getMonth();
        this.showCalendar();
    }

    async showCalendar() {
        if (this.isAdmin) {
            await this.showMonthlyView();
        } else {
            await this.showYearlyView();
        }
    }

    async showMonthlyView() {
        document.getElementById('monthly-calendar').style.display = 'block';
        document.getElementById('yearly-calendar').style.display = 'none';
        document.getElementById('calendar-summary').style.display = 'none';

        // Mettre à jour le titre
        const title = `${this.monthNames[this.currentMonth]} ${this.currentYear}`;
        document.getElementById('calendar-title').textContent = title;

        // Charger les événements du mois
        try {
            const response = await fetch(`${CONFIG.API_BASE_URL}/calendar/admin?year=${this.currentYear}&month=${this.currentMonth + 1}`, {
                headers: {
                    'Authorization': `Bearer ${authToken}`
                }
            });

            if (!response.ok) {
                throw new Error('Erreur lors du chargement du calendrier');
            }

            const data = await response.json();
            this.events = Array.isArray(data) ? data : [];
            this.renderMonthlyCalendar();
        } catch (error) {
            console.error('Erreur:', error);
            showAlert('Erreur lors du chargement du calendrier', 'error');
        }
    }

    async showYearlyView() {
        document.getElementById('monthly-calendar').style.display = 'none';
        document.getElementById('yearly-calendar').style.display = 'block';
        document.getElementById('calendar-summary').style.display = 'block';

        // Mettre à jour le titre
        document.getElementById('calendar-title').textContent = this.currentYear.toString();

        // Charger les événements de l'année et le résumé
        try {
            const [eventsResponse, summaryResponse] = await Promise.all([
                fetch(`${CONFIG.API_BASE_URL}/calendar/user?year=${this.currentYear}`, {
                    headers: {
                        'Authorization': `Bearer ${authToken}`
                    }
                }),
                fetch(`${CONFIG.API_BASE_URL}/calendar/summary?year=${this.currentYear}`, {
                    headers: {
                        'Authorization': `Bearer ${authToken}`
                    }
                })
            ]);

            if (!eventsResponse.ok || !summaryResponse.ok) {
                throw new Error('Erreur lors du chargement du calendrier');
            }

            const data = await eventsResponse.json();
            this.events = Array.isArray(data) ? data : [];
            const summary = await summaryResponse.json();
            
            // Afficher le résumé
            const summaryText = `${summary.used_leave_days}/${summary.total_leave_days} jours utilisés - ${summary.remaining_leave_days} jours restants`;
            document.getElementById('summary-text').textContent = summaryText;

            this.renderYearlyCalendar();
        } catch (error) {
            console.error('Erreur:', error);
            showAlert('Erreur lors du chargement du calendrier', 'error');
        }
    }

    renderMonthlyCalendar() {
        const daysContainer = document.getElementById('calendar-days');
        daysContainer.innerHTML = '';

        const firstDay = new Date(this.currentYear, this.currentMonth, 1);
        const lastDay = new Date(this.currentYear, this.currentMonth + 1, 0);
        const daysInMonth = lastDay.getDate();
        const startingDayOfWeek = (firstDay.getDay() + 6) % 7; // Lundi = 0

        // Ajouter les jours vides du mois précédent
        for (let i = 0; i < startingDayOfWeek; i++) {
            const dayDiv = document.createElement('div');
            dayDiv.className = 'calendar-day empty';
            daysContainer.appendChild(dayDiv);
        }

        // Ajouter les jours du mois
        for (let day = 1; day <= daysInMonth; day++) {
            const dayDiv = document.createElement('div');
            dayDiv.className = 'calendar-day';
            
            const dayNumber = document.createElement('div');
            dayNumber.className = 'day-number';
            dayNumber.textContent = day;
            dayDiv.appendChild(dayNumber);

            // Vérifier si c'est aujourd'hui
            const today = new Date();
            if (today.getFullYear() === this.currentYear && 
                today.getMonth() === this.currentMonth && 
                today.getDate() === day) {
                dayDiv.classList.add('today');
            }

            // Ajouter les événements du jour
            const dayDate = new Date(this.currentYear, this.currentMonth, day);
            const dayEvents = this.getEventsForDate(dayDate);
            
            if (dayEvents.length > 0) {
                const eventsContainer = document.createElement('div');
                eventsContainer.className = 'day-events';
                
                dayEvents.forEach(event => {
                    const eventDiv = document.createElement('div');
                    // Ajouter une classe spéciale pour les déclarations de maladie
                    let eventClass = `event event-${event.status} event-${event.type}`;
                    if (event.event_source === 'sickness_declaration') {
                        eventClass += ' event-sickness-declaration';
                    }
                    eventDiv.className = eventClass;
                    eventDiv.textContent = this.truncateText(event.title, 20);
                    eventDiv.addEventListener('click', (e) => {
                        e.stopPropagation();
                        this.showEventModal(event);
                    });
                    eventsContainer.appendChild(eventDiv);
                });
                
                dayDiv.appendChild(eventsContainer);
            }

            daysContainer.appendChild(dayDiv);
        }
    }

    renderYearlyCalendar() {
        const yearGrid = document.querySelector('.year-grid');
        if (!yearGrid) {
            console.error('Element .year-grid not found!');
            return;
        }
        yearGrid.innerHTML = '';
        


        for (let month = 0; month < 12; month++) {
            const monthDiv = document.createElement('div');
            monthDiv.className = 'mini-month';
            
            const monthHeader = document.createElement('div');
            monthHeader.className = 'mini-month-header';
            monthHeader.textContent = this.monthNames[month];
            monthDiv.appendChild(monthHeader);

            const monthGrid = document.createElement('div');
            monthGrid.className = 'mini-month-grid';
            
            // En-têtes des jours de la semaine (version courte)
            const weekdays = ['L', 'M', 'M', 'J', 'V', 'S', 'D'];
            weekdays.forEach(day => {
                const weekdayDiv = document.createElement('div');
                weekdayDiv.className = 'mini-weekday';
                weekdayDiv.textContent = day;
                monthGrid.appendChild(weekdayDiv);
            });

            // Jours du mois
            const firstDay = new Date(this.currentYear, month, 1);
            const lastDay = new Date(this.currentYear, month + 1, 0);
            const daysInMonth = lastDay.getDate();
            const startingDayOfWeek = (firstDay.getDay() + 6) % 7;

            // Jours vides
            for (let i = 0; i < startingDayOfWeek; i++) {
                const emptyDiv = document.createElement('div');
                emptyDiv.className = 'mini-day empty';
                monthGrid.appendChild(emptyDiv);
            }

            // Jours du mois
            for (let day = 1; day <= daysInMonth; day++) {
                const dayDiv = document.createElement('div');
                dayDiv.className = 'mini-day';
                dayDiv.textContent = day;

                // Vérifier si c'est aujourd'hui
                const today = new Date();
                if (today.getFullYear() === this.currentYear && 
                    today.getMonth() === month && 
                    today.getDate() === day) {
                    dayDiv.classList.add('today');
                }

                // Vérifier s'il y a des événements
                const dayDate = new Date(this.currentYear, month, day);
                const dayEvents = this.getEventsForDate(dayDate);
                
                if (dayEvents.length > 0) {
                    dayDiv.classList.add('has-events');
                    // Ajouter une classe pour le type d'événement principal
                    const primaryEvent = dayEvents[0];
                    dayDiv.classList.add(`event-${primaryEvent.status}`, `event-${primaryEvent.type}`);
                    
                    // Ajouter une classe spéciale si c'est une déclaration de maladie
                    if (primaryEvent.event_source === 'sickness_declaration') {
                        dayDiv.classList.add('has-sickness-declaration');
                    }
                    
                    // Ajouter un tooltip ou gérer le clic
                    dayDiv.title = dayEvents.map(e => e.title).join('\n');
                    dayDiv.addEventListener('click', () => {
                        if (dayEvents.length === 1) {
                            this.showEventModal(dayEvents[0]);
                        } else {
                            // Afficher une liste des événements
                            this.showMultipleEventsModal(dayEvents, dayDate);
                        }
                    });
                }

                monthGrid.appendChild(dayDiv);
            }

            monthDiv.appendChild(monthGrid);
            yearGrid.appendChild(monthDiv);
        }
    }

    getEventsForDate(date) {
        if (!Array.isArray(this.events)) return [];
        return this.events.filter(event => {
            // Utiliser les dates en format string pour éviter les problèmes de fuseau horaire
            const dateStr = date.toISOString().split('T')[0]; // Format YYYY-MM-DD
            const eventStart = event.start; // Déjà en format YYYY-MM-DD
            const eventEnd = event.end; // Déjà en format YYYY-MM-DD
            
            return dateStr >= eventStart && dateStr <= eventEnd;
        });
    }

    truncateText(text, maxLength) {
        return text.length > maxLength ? text.substring(0, maxLength) + '...' : text;
    }

    showEventModal(event) {
        document.getElementById('event-title').textContent = event.title;
        document.getElementById('event-user').textContent = event.user_name;
        
        // Différencier l'affichage selon le type d'événement
        if (event.event_source === 'sickness_declaration') {
            document.getElementById('event-type').textContent = 'Déclaration de maladie';
        } else {
            document.getElementById('event-type').textContent = event.type === 'vacances' ? 'Vacances' : 'Maladie';
        }
        
        const startDate = new Date(event.start).toLocaleDateString('fr-FR');
        const endDate = new Date(event.end).toLocaleDateString('fr-FR');
        document.getElementById('event-dates').textContent = startDate === endDate ? startDate : `${startDate} - ${endDate}`;
        
        // Affichage du statut différent pour les déclarations de maladie
        if (event.event_source === 'sickness_declaration') {
            const emailSent = event.title.includes('✉️');
            document.getElementById('event-status').textContent = emailSent ? 'Email envoyé' : 'Email non envoyé';
        } else {
            const statusText = {
                'en_attente': 'En attente',
                'approuve': 'Approuvé',
                'refuse': 'Refusé'
            }[event.status] || event.status;
            document.getElementById('event-status').textContent = statusText;
        }
        
        const reasonRow = document.getElementById('event-reason-row');
        if (event.reason) {
            document.getElementById('event-reason').textContent = event.reason;
            reasonRow.style.display = 'block';
        } else {
            reasonRow.style.display = 'none';
        }

        document.getElementById('event-modal').style.display = 'flex';

        // Activer les actions admin pour les demandes d'absence uniquement
        const isAdmin = currentUser && currentUser.role === 'admin';
        const actionsRow = document.querySelector('#event-modal .admin-only');
        const editBtn = document.getElementById('event-edit-btn');
        const deleteBtn = document.getElementById('event-delete-btn');
        if (actionsRow && editBtn && deleteBtn) {
            if (isAdmin && event.event_source === 'absence_request') {
                actionsRow.style.display = 'flex';
                editBtn.onclick = () => this.openEditDialog(event);
                deleteBtn.onclick = () => this.confirmDelete(event);
            } else {
                actionsRow.style.display = 'none';
                editBtn.onclick = null;
                deleteBtn.onclick = null;
            }
        }
    }

    

    showMultipleEventsModal(events, date) {
        // Pour simplifier, on affiche juste le premier événement
        // Dans une vraie app, on pourrait créer une modal spéciale pour plusieurs événements
        this.showEventModal(events[0]);
    }

    async confirmDelete(event) {
        if (!confirm('Supprimer cette absence ?')) return;
        try {
            await apiCall(`/absence-requests/admin/${event.id}`, { method: 'DELETE' });
            showAlert('Absence supprimée');
            closeEventModal();
            await this.showCalendar();
            loadAllRequests && loadAllRequests();
        } catch (e) {
            showAlert(e.message || 'Erreur suppression', 'error');
        }
    }

    openEditDialog(event) {
        const newStart = prompt('Nouvelle date de début (YYYY-MM-DD):', event.start);
        if (!newStart) return;
        const newEnd = prompt('Nouvelle date de fin (YYYY-MM-DD):', event.end);
        if (!newEnd) return;
        const newReason = prompt('Raison (optionnel):', event.reason || '');
        const statusMap = { 'en_attente': 'en_attente', 'approuve': 'approuve', 'refuse': 'refuse' };
        const newStatus = prompt('Statut (en_attente/approuve/refuse):', event.status);
        const normalizedStatus = statusMap[newStatus] ? newStatus : undefined;
        this.updateAbsence(event.id, {
            start_date: newStart,
            end_date: newEnd,
            reason: newReason || null,
            status: normalizedStatus
        });
    }

    async updateAbsence(id, payload) {
        try {
            await apiCall(`/absence-requests/admin/${id}`, {
                method: 'PUT',
                body: JSON.stringify(payload)
            });
            showAlert('Absence mise à jour');
            closeEventModal();
            await this.showCalendar();
            loadAllRequests && loadAllRequests();
        } catch (e) {
            showAlert(e.message || 'Erreur mise à jour', 'error');
        }
    }
}

function closeEventModal() {
    document.getElementById('event-modal').style.display = 'none';
}

// Initialiser le calendrier quand la section est affichée
let calendar = null;

async function showCalendarSection() {
    const calendarSection = document.getElementById('calendar-section');
    if (!calendarSection) {
        console.error('calendar-section element not found!');
        return;
    }
    
    calendarSection.style.display = 'block';
    
    if (!calendar) {
        calendar = new Calendar();
        await calendar.init();
    } else {
        await calendar.showCalendar();
    }
}